lib/search/embeddings/embedding.ex

```diff
@@ -0,0 +1,137 @@
+    %{
+      model: model_repo,
+      embedding_size: embedding_size,
+      serving_opts: serving_opts,
+      load_model_opts: load_model_opts,
+      load_tokenizer_opts: load_tokenizer_opts
+    } =
+      opts
+      |> Keyword.validate!([
+        :model,
+        :embedding_size,
+        serving_opts: [],
+        load_model_opts: [],
+        load_tokenizer_opts: []
+      ])
+      |> Map.new()
+
+    batch_size =
+      serving_opts
+      |> Keyword.get(:compile, [])
+      |> Keyword.get(:batch_size)
+
+    table_name =
+      quote do
+        "#{Macro.underscore(__MODULE__)}_embeddings"
+      end
+
+    text_batching =
+      if batch_size do
+        quote do
+          Stream.map(& &1.text)
+          |> Stream.chunk_every(unquote(batch_size))
+        end
+      else
+        quote do
+          Stream.map(& &1.text)
+        end
+      end
+
+    quote do
+      alias Search.{Packages, Repo}
+      use Ecto.Schema
+      import Ecto.{Changeset, Query}
+      alias Pgvector.Ecto.Vector
+
+      def table_name do
+        unquote(table_name)
+      end
+
+      schema unquote(table_name) do
+        field :embedding, Vector
+        belongs_to :doc_fragment, Packages.DocFragment
+
+        timestamps(type: :utc_datetime)
+      end
+
+      def changeset(embedding, attrs) do
+        embedding
+        |> cast(attrs, [:embedding])
+        |> cast_assoc(:doc_fragment, required: true)
+        |> validate_required([:embedding])
+      end
+
+      def embedding_size, do: unquote(embedding_size)
+      def batch_size, do: unquote(batch_size)
+
+      def child_spec(opts) do
+        opts
+        |> Keyword.merge(serving: load_model())
+        |> Nx.Serving.child_spec()
+      end
+
+      @doc """
+      Embeds any doc fragments which do not have an embedding yet. Recieves an optional callback,
+      which is called to notify about the embedding progress with the tuple {total, done} as its argument.
+      """
+      def embed(progress_callback \\ &Function.identity/1) do
+        fragments =
+          from f in Packages.DocFragment,
+            left_join: e in __MODULE__,
+            on: e.doc_fragment_id == f.id,
+            where: is_nil(e)
+
+        fragments = Repo.all(fragments)
+
+        if fragments == [] do
+          {:ok, []}
+        else
+          fragments_count = length(fragments)
+
+          fragment_texts = fragments |> unquote(text_batching)
+
+          embeddings =
+            fragment_texts
+            |> Stream.with_index(1)
+            |> Stream.flat_map(fn {text, index} ->
+              embeddings = Nx.Serving.batched_run(__MODULE__, text)
+
+              progress_callback.({fragments_count, min(index * batch_size(), fragments_count)})
+
+              case embeddings do
+                %{embedding: embedding} ->
+                  embedding
+
+                _ when is_list(embeddings) ->
+                  Stream.map(embeddings, & &1.embedding)
+              end
```

Comment:
To avoid two code paths, for the singular case you could wrap each text in a list. That said, I would just require the `batch_size` to be specified in the config (with an aggressive `Keyword.fetch!/2`).